addons/sale/models/sale_order.py

```diff
@@ -1793,6 +1793,50 @@ def _generate_downpayment_invoices(self):
 
         return generated_invoices
 
+    def _get_product_catalog_order_line_info(self, product_ids, **kwargs):
+        """ Returns products information to be shown in the catalog.
+        :param list product_ids: The products currently displayed in the product catalog, as a list
+                                 of `product.product` ids.
+        :rtype: dict
+        :return: A dict with the following structure:
+            {
+                'productId': int
+                'quantity': float (optional)
+                'price': float
+                'readOnly': bool (optional)
+            }
+        """
+        sale_order_line_info = {}
+        for product, lines in groupby(
+            self.order_line.filtered(lambda line: not line.display_type),
+            lambda line: line.product_id
+        ):
+            if(product.id not in product_ids):
+                continue
+
+            sale_order_line_info[product.id] = self.env['sale.order.line'].browse(
+                line.id for line in lines # groupby gives a list and not a recordset
+            )._get_catalog_info(**kwargs)
+
+            product_ids.remove(product.id)
+
+        default_data = self.env['sale.order.line']._get_catalog_info(**kwargs)
+        default_data['readOnly'] = self._is_readonly()
+
+        sale_order_line_info.update({
+            product_id: {
+                **default_data,
+                'price': price,
+            }
+            for product_id, price in self.pricelist_id._get_products_price(
+                quantity=1.0,
+                products=self.env['product.product'].browse(product_ids),
+                currency=self.currency_id,
+                date=self.date_order,
+            ).items()
+        })
+        return sale_order_line_info
```

Comment:
Since the expected API is defined in the `catalog` controller of `product`, shouldn't we factorize the shared API in a mixin clearly defining the expected methods, their return value, ... (and maybe factorize some code between sale & purchase) ?

(You could even include `action_add_from_catalog` in the mixin)

Reply:
@Feyensv mixin done but it still a bit a mess.
Some parts can't be correctly factorized (without doing crazy things at least) since we can't assume the line's model and the field name.
Maybe it possible to make it more clean but since I'm not used to mixin creation and we are in the rush, it's a bit complicated